### PR TITLE
Use go-connections, and TLS flags for notary client

### LINKED ISSUE #452: Allow notary to load locally trusted TLS certificates using Docker's mechanism
Currently to configure the ability of a notary-client to connect to a remote notary-server that has a non-public CA signed certificate, we need a configuration file change:

```
"remote_server": {
	"url": "https://107.170.203.81",
	"root_ca": "/home/diogo/.docker/tls/107.170.203.81/ca.crt"
}
```

Docker uses this package: https://github.com/docker/go-connections

And the relevant code:

```
const (
	defaultTrustKeyFile = "key.json"
	defaultCaFile       = "ca.pem"
	defaultKeyFile      = "key.pem"
	defaultCertFile     = "cert.pem"
)
...
dockerCertPath  = os.Getenv("DOCKER_CERT_PATH")
...
if dockerCertPath == "" {
	dockerCertPath = cliconfig.ConfigDir()
}
...
	cmd.StringVar(&tlsOptions.CAFile, []string{"-tlscacert"}, filepath.Join(dockerCertPath, defaultCaFile), "Trust certs signed only by this CA")
	cmd.StringVar(&tlsOptions.CertFile, []string{"-tlscert"}, filepath.Join(dockerCertPath, defaultCertFile), "Path to TLS certificate file")
	cmd.StringVar(&tlsOptions.KeyFile, []string{"-tlskey"}, filepath.Join(dockerCertPath, defaultKeyFile), "Path to TLS key file")
...
```

### PR DESCRIPTION
**Pull request by cyli**
This deletes the TLS configuration we had before, and just uses go-connections.

It also adds the following flags for the notary client CLI:

`--tlscacert`, `--tlscert`, and `--tlskey`, making this match the docker client TLS flags.

This also enables the notary client to support mutual authentication with the server.

Fixes #452

### CODE REVIEW

File: cmd/notary/main.go [hunk 1]
```diff
@@ -118,8 +142,9 @@ func (n *notaryCommander) GetCommand() *cobra.Command {
 		Use:           "notary",
 		Short:         "Notary allows the creation of trusted collections.",
 		Long:          "Notary allows the creation and management of collections of signed targets, allowing the signing and validation of arbitrary content.",
-		SilenceUsage:  true,
-		SilenceErrors: true,
+		SilenceUsage:  true, // we don't want to print out usage for EVERY error
+		SilenceErrors: true, // we do our own error reporting with fatalf
```
Comment (riyazdf) [inline cmd/notary/main.go]: thanks for documenting these flags :+1:

File: cmd/notary/main_test.go [hunk 2]
```diff
@@ -162,8 +167,203 @@ func TestInsufficientArgumentsReturnsErrorAndPrintsUsage(t *testing.T) {
 			[]string{"-c", filepath.Join(tempdir, "idonotexist.json"), "-d", tempdir}, arglist...))
 		err = cmd.Execute()
 
-		require.NotContains(t, err.Error(), "error opening config file", "running %s", invalid)
+		require.NotContains(t, err.Error(), "error opening config file", "running `notary %s`", invalid)
 		// it's a usage error, so the usage is printed
-		require.Contains(t, b.String(), "Usage:", "expected usage when running %s", invalid)
+		require.Contains(t, b.String(), "Usage:", "expected usage when running `notary %s`", invalid)
```
Comment (riyazdf) [inline cmd/notary/main_test.go]: thanks for fixing these too :+1:

File: cmd/notary/main_test.go [hunk 3]
```diff
@@ -162,8 +167,203 @@ func TestInsufficientArgumentsReturnsErrorAndPrintsUsage(t *testing.T) {
...
+		require.Contains(t, b.String(), "Usage:", "expected usage when running `notary %s`", bareCommand)
+	}
+}
+
+type recordingMetaStore struct {
+	gotten []string
+	storage.MemStorage
+}
+
+// GetCurrent records what metadata was requested
```
Comment (riyazdf) [inline cmd/notary/main_test.go]: nit: maybe the comment should mention that it also gets the metadata from the underlying metastore? (same for `GetChecksum`)

File: utils/configuration_test.go [hunk 4]
```diff
@@ -234,28 +244,27 @@ func TestParseStorageWithEnvironmentVariables(t *testing.T) {
...
 	for _, configJSON := range invalids {
 		_, err := ParseServerTLS(configure(configJSON), true)
 		assert.Error(t, err)
-		assert.Contains(t, err.Error(),
-			"both the TLS certificate and key are mandatory")
+		assert.Contains(t, err.Error(), "no such file or directory")
 	}
 }
 
 // If TLS is not and the cert/key are partially provided, an error is returned
```
Comment (riyazdf) [inline utils/configuration_test.go]: nit: while you're in this file, can we update this comment to start with `// If TLS is not required`?

**Comment by riyazdf** [comment]
lots of great tests in here, LGTM

File: utils/configuration_test.go (hunk 4 above)
Comment (cyli) [inline utils/configuration_test.go]: Good spot, will do!  :D

File: cmd/notary/main.go (hunk 1 above)
Comment (HuKeping) [inline cmd/notary/main.go]: :+1:  for this

File: cmd/notary/main_test.go [hunk 5]
```diff
@@ -162,8 +167,205 @@ func TestInsufficientArgumentsReturnsErrorAndPrintsUsage(t *testing.T) {
...
+
+// GetCurrent gets the metadata from the underlying MetaStore, but also records
+// that the metadata was requested
+func (r *recordingMetaStore) GetCurrent(gun, role string) (data []byte, err error) {
+	r.gotten = append(r.gotten, fmt.Sprintf("%s.%s", gun, role))
+	return r.MemStorage.GetCurrent(gun, role)
+}
+
+// GetChecksum gets the metadata from the underlying MetaStore, but also records
+// that the metadata was requested
```
Comment (HuKeping) [inline cmd/notary/main_test.go]: The same comments as above in `GetCurrent()`

Comment (HuKeping) [inline cmd/notary/main_test.go]: The same comments as above in `GetCurrent()`

Comment (HuKeping) [inline cmd/notary/main_test.go]: Seems it's intended :santa:

**Comment by diogomonica** [comment]
LGTM

**Pull request merged**